consensus/istanbul/validator/weighted_random_test.go

```diff
@@ -441,32 +443,92 @@ func TestWeightedCouncil_SetSubGroupSize(t *testing.T) {
+			{2, 3, 6, 8, 10, 9, 1, 11, 5, 0, 4, 7},
+			{3, 4, 10, 5, 8, 0, 7, 9, 12, 6, 1, 11},
+			{4, 5, 6, 0, 7, 1, 3, 12, 2, 8, 9, 11},
+			{5, 6, 4, 9, 12, 7, 0, 3, 8, 2, 1, 11},
+			{6, 7, 2, 3, 8, 9, 11, 12, 5, 1, 4, 0},
+			{7, 8, 3, 11, 5, 10, 0, 1, 2, 6, 9, 12},
+			{8, 9, 5, 7, 11, 3, 1, 0, 10, 6, 4, 12},
+			{9, 10, 3, 7, 5, 6, 2, 0, 12, 8, 11, 1},
+			{10, 11, 7, 1, 0, 9, 8, 6, 12, 5, 2, 4},
+			{11, 12, 4, 8, 1, 6, 0, 3, 9, 10, 2, 7},
+			{12, 0, 7, 2, 4, 1, 6, 10, 9, 11, 8, 3},
+			{0, 1, 9, 8, 2, 3, 10, 5, 7, 12, 4, 6},
+			{1, 2, 11, 10, 6, 8, 7, 4, 9, 12, 0, 5},
+			{2, 3, 4, 8, 6, 5, 11, 1, 12, 0, 9, 10},
+		}
+	)
 
-	valSet.SetBlockNum(1)
-	view := &istanbul.View{
-		Sequence: new(big.Int).SetInt64(1),
-		Round:    new(big.Int).SetInt64(0),
+	getExpectSubList := func(indices []int) []istanbul.Validator {
+		var expectSubList []istanbul.Validator
+		for _, index := range indices {
+			expectSubList = append(expectSubList, validators[index])
+		}
+		return expectSubList
 	}
 
-	for i := 2; i < len(validators); i++ {
-		testSubSetLen := uint64(i)
-		valSet.SetSubGroupSize(testSubSetLen)
+	// set hardForkBlockNumberConfig and view values
+	fork.SetHardForkBlockNumberConfig(&params.ChainConfig{IstanbulCompatibleBlock: istanbul.TestBlockHF})
 
-		testSubList := valSet.SubListWithProposer(crypto.Keccak256Hash([]byte("This is a test")), valSet.GetProposer().Address(), view)
-		resultSubListLen := len(testSubList)
+	// SubsetLen test: various subset length test
+	valSet.SetBlockNum(1)
+	for testSubsetLen := 2; testSubsetLen < len(validators); testSubsetLen++ {
+		// set committee size and calculate proposer
+		valSet.SetSubGroupSize(uint64(testSubsetLen))
+		valSet.CalcProposer(valSet.GetProposer().Address(), uint64(0))
 
-		if int(testSubSetLen) != resultSubListLen {
-			t.Errorf("SubGroupSize should be %v but gotten SubGroupSize is %v", testSubSetLen, resultSubListLen)
-		}
+		// get committee list
+		expectSubList := getExpectSubList(expectIndexOfSubsetLenTest[0:testSubsetLen])
 
-		for j := 0; j < resultSubListLen; j++ {
-			_, validator := valSet.getByAddress(testSubList[j].Address())
-			if validator == nil {
-				t.Errorf("validator in subGroup is not an element of weightedCouncil. validaotr address %v ", testSubList[j].Address())
-			}
-		}
+		// compare the subList of valSet with expected committee list
+		viewBeforeHF := &istanbul.View{Sequence: istanbul.TestBlockBeforeHF, Round: big.NewInt(int64(0))}
+		viewAfterHF := &istanbul.View{Sequence: istanbul.TestBlockAfterHF, Round: big.NewInt(int64(0))}
+		assert.Equal(t, expectSubList, valSet.SubList(prevHash, viewBeforeHF), "test Subset length: %d(before istanbulCompatible)", testSubsetLen)
+		assert.Equal(t, expectSubList, valSet.SubList(prevHash, viewAfterHF), "test subset length: %d(after istanbulCompatible)", testSubsetLen)
+	}
+
+	// Round test: various round test
+	valSet.SetBlockNum(1)
+	valSet.SetSubGroupSize(uint64(len(validators) - 1))
+	for round := 0; round < 15; round++ {
```

Comment:
`15` can be changed to `len(expectIndexOfRoundTestBeforeIstanbulCompatible)`. 
It may reduce mistakes when a new test array is added. 
It would be better if this code check whether `len(expectIndexOfRoundTestBeforeIstanbulCompatible) == len(expectIndexOfRoundTestAfterIstanbulCompatible)`

Reply:
reflected at https://github.com/klaytn/klaytn/pull/997/commits/d2f24939698fe688dba58022ea14ff1aacb17d5e, thanks!